chess-AI/experimental-work/mcts.py

```diff
@@ -1,73 +1,161 @@
-#https://web.stanford.edu/~surag/posts/alphazero.html
+from math import sqrt
+import numpy as np
+from state_representation import get_cnn_input
 
 class Mcts:
-    def search(s,game,nnet):
 
-        if game.gameEnded(s): return  -game.gameReward(s) # returns the negative of the game reward, why negative?
+    def __init__(self):
+        self.states_visited = [] # stores fen strings of states visited
+        self.n_values = {} # stores all the N values
+        self.q_values = {} # stores all the Q values
+        self.exploration = 1 # sets exploration rate
+        self.p_values = {} # stores the probabilities computed by NN
 
-        if s not in visited # if a state has not been visited then you must find the predictions made by the model
-            visited.add(s) # will mark as visisted
-            P[s],v = nnet.predict(s) # this is like the rollout phase but using the nueral network
-            return -value
-        
-        max_u, best_a = -float("inf"), -1 
+    def find_best_move(self,legal_moves,fen_string):
+
+        best_u = -float("inf")
+        best_move = -1
+
+        # Iterate through all legal moves at the state
+
+        for move in legal_moves:
+
+            # Get move in uci form
+            uci_move = move.uci()
+
+            # Check to see if the state has been added to Q
+            if fen_string in self.q_values:
+                if uci_move in self.q_values[fen_string]:
+                    q_value = self.q_values[fen_string][uci_move]
+                else:
+                    self.q_values[fen_string][uci_move] = 0
+                    q_value = 0
+            else:
+                self.q_values[fen_string]={}
+                self.q_values[fen_string][uci_move] = 0
+                q_value = 0
+
+
+            # Check to see if state has been added to N
+            if fen_string in self.n_values:
+                if uci_move in self.n_values[fen_string]:
+                    n_value = self.n_values[fen_string][uci_move]
+                else:
+                    self.n_values[fen_string][uci_move] = 1
+                    n_value = float("inf")
+            else:
+                self.n_values[fen_string]={}
+                self.n_values[fen_string][uci_move] = 1
+                n_value = float("inf")
+
+            """
+            NOTE: still unsure about the n values
+            not sure if I should set the n values to infinite if a node
+            has not been visited
+            Also do not know what N should be initial
+            """
```

Comment:
This should be a `# TODO` type comment. Triple quote strings should only be used for docstrings.

This could instead be
```python
# TODO: Initial N value for unvisited nodes may need to be updated. Revisit this.
```